src/roswire/common/source.py

```diff
@@ -134,6 +107,170 @@ def for_app_instance(
+        join = os.path.join(package.path, new_env['cwd'])
+        cmakelists_path = os.path.join(join, 'CMakeLists.txt')
+        logger.debug(f"Processing {cmakelists_path}")
+        included_package_info = self.process_cmake_contents(
+            self._files.read(cmakelists_path),
+            package,
+            new_env,
+        )
+        executables = {
+            **executables,
+            **{s: included_package_info.executables[s]
+               for s in included_package_info.executables}
+        }
+        return executables
+
+    def __process_add_executable(
+        self,
+        args: t.List[str],
+        cmake_env: t.Dict[str, str],
+        executables: t.Dict[str, ExecutableInfo],
+        package: Package
+    ) -> None:
+        name = args[0]
+        sources: t.Set[str] = set()
+        for source in args[1:]:
+            if 'cwd' in cmake_env:
+                sources.add(os.path.join(cmake_env['cwd'], source))
+            else:
+                sources.add(source)
+        logger.debug(f"Adding C++ sources for {name}")
+        executables[name] = ExecutableInfo(
+            name=name,
+            language=SourceLanguage.CXX,
+            kind=ExecutableKind.NODE,
+            sources=sources,
+            restrict_to_paths=self.package_paths(package))
+
+    def __process_add_library(
+        self,
+        args: t.List[str],
+        cmake_env: t.Dict[str, str],
+        executables: t.Dict[str, ExecutableInfo],
+        package: Package
+    ) -> None:
+        name = args[0]
+        if 'cwd' in cmake_env:
+            sources = {os.path.join(cmake_env['cwd'], s) for s in args[1:]}
+        else:
+            sources = set(args[1:])
+        logger.debug(f"Adding C++ library {name}")
+        executables[name] = ExecutableInfo(
+            name,
+            SourceLanguage.CXX,
+            ExecutableKind.LIBRARY,
+            sources,
+            self.package_paths(package))
+
+    def __process_python_executables(
+        self,
+        args: t.List[str],
+        cmake_env: t.Dict[str, str],
+        executables: t.Dict[str, ExecutableInfo]
+    ) -> None:
+        opts, args = cmake_argparse(
+            args,
+            {"PROGRAMS": "*", "DESTINATION": "*"}
+        )
+        if 'PROGRAMS' in opts:
+            for i in range(len(opts['PROGRAMS'])):
+                # http://docs.ros.org/en/jade/api/catkin/html/howto/format2/installing_python.html  # noqa: F401, E501
+                # Convention is that ros python nodes are in nodes/ directory.
+                # All others are in scripts/. So just include python installs
+                # that are in nodes/
+                program = opts['PROGRAMS'][i]
+                if program.startswith("nodes/"):
+                    name = Path(program[0]).stem
+                    sources = set(program)
+                    if 'cwd' in cmake_env:
+                        sources = set(os.path.join(cmake_env['cwd'], program))
```

Comment:
If I remember correctly, this will produce a set of single-character strings obtained from iterating over the result of `os.path.join`.

Reply:
Fixed